src/compiler/utilities.ts

```diff
@@ -3713,10 +3717,14 @@ namespace ts {
         return (kind >= SyntaxKind.FirstTypeNode && kind <= SyntaxKind.LastTypeNode)
             || kind === SyntaxKind.AnyKeyword
             || kind === SyntaxKind.NumberKeyword
+            || kind === SyntaxKind.ObjectKeyword
             || kind === SyntaxKind.BooleanKeyword
             || kind === SyntaxKind.StringKeyword
             || kind === SyntaxKind.SymbolKeyword
+            || kind === SyntaxKind.ThisKeyword

```

Comment:
resolved offline: we decided that conflating `this` as a type node and expression isn't a problem for now because it has no children and we would only rely on the distinction for now when aggregating modifier flags from children.

We may want to introduce a distinction as we have with `StringLiteral` and `LiteralType` in the future.